torch/onnx/_internal/fx/passes/fx_to_onnxscript.py

```diff
@@ -0,0 +1,510 @@
+            current_attr = getattr(current_attr, sub_attr_name)
+
+        input_ = onnxscript_graph.add_input(
+            input_name=node.name, input_value=current_attr
+        )
+        assert isinstance(input_, graph_building.TorchScriptTensor)
+        assert isinstance(input_, onnxscript.tensor.Tensor)
+        fx_name_to_onnxscipt_value[node.name] = input_
+        onnxscript_value_name_to_real_tensor[input_.name] = current_attr  # type: ignore[assignment]
+    else:
+        # TODO(wechi): Support get_attr, call_module, call_method.
+        raise RuntimeError(f"Found node type not defined in torch.fx: {node.op}")
+
+
+@diagnostics.diagnose_call(diagnostics.rules.atenlib_fx_to_onnx)
+def export_fx_to_onnxscript(
+    fx_module_with_metadata: torch.fx.GraphModule, options: options.ExportOptions
+):
+
+    # Initialize the ONNX graph
+    onnxscript_graph = graph_building.TorchScriptGraph()
+    tracer = graph_building.TorchScriptTracingEvaluator(onnxscript_graph)
+
+    # In the following loop, a TorchScript graph is created to
+    # represent the input FX graph with ONNX symbols (e.g., onnx::add).
+    # To connect the values to nodes in the TorchScript graph, we maintian
+    # fx_name_to_onnxscipt_value. Basically, we want to translate
+    #   fx_tensor_x (type: torch.fx.Node) -> fx_node_1 -> fx_tensor_y (type: torch.fx.Node)
+    # to
+    #   fx_name_to_onnxscipt_value[fx_tensor_x.name] -> onnx_node_1 -> fx_name_to_onnxscipt_value[fx_tensor_y.name]
+    fx_name_to_onnxscipt_value: Dict[
+        str, Union[torch._C.Value, Tuple[torch._C.Value, ...]]
+    ] = {}
+    # Similar to fx_name_to_onnxscipt_value, we need a mapping fo real tensors (usually tensor parameters
+    # in nn.Module). Note that TorchScript's cannot store real tensors; TorchScript values are all
+    # symbolic. This is passed into ONNX ModelProto as the initializers.
+    onnxscript_value_name_to_real_tensor: Dict[
+        str, Union[torch.Tensor, Tuple[torch._C.Value, ...]]
+    ] = {}
+    for node in fx_module_with_metadata.graph.nodes:
+        _export_fx_node_to_onnxscript(
+            node,
+            onnxscript_graph,
+            fx_name_to_onnxscipt_value,
+            onnxscript_value_name_to_real_tensor,
+            tracer,
+            fx_module_with_metadata,
+            options,
+        )
+
+    # Apply TorchScript's type promotion code.
+    # Ideally, we should implement our type promotion but
+    # to save time, we just reuse.
+    onnxscript_graph.apply(
+        torch._C._jit_pass_onnx_scalar_type_analysis,
+        lowprecision_cast=True,
+        opset_version=options.opset_version,
+    )
+
+    return onnxscript_graph, onnxscript_value_name_to_real_tensor
+
+
+@_beartype.beartype
+def _validate_op_between_ort_torch(
+    node: torch.fx.Node,
+    symbolic_fn: Union[onnxscript.OnnxFunction, Callable],
+    torch_args: tuple,
+    torch_kwargs: dict,
+):
+    """Validate the op between ONNX Runtime and PyTorch."""
```

Comment:
What does this function really do? The doc string is basically the same as its name.

Reply:
This PR only splits code from one file to several.

Reply:
This function is trying to validate op-by-op results with runtime shape generated inputs to provide op-level debugging. I have a plan to refactor this once we have settled the shape inference of graph (Currently, we don't have shape in proto and I haven't figured out how dynamic axes would affect the shape usage in this function, but this is default off).